### PR TITLE
fw_tecs: Support tighter altitude tracking during low-height flight

### PR DESCRIPTION
**Pull request by oravla5**
### Solved Problem
Slow TECS  reaction to altitude errors during low-height flight could lead to collisions with trees or other high obstacles in the flight path.

### Solution
The new parameter `FW_T_THR_LOW_HGT` allows to define a height threshold under which the TECS altitude tracking becomes tighter according to the already existing parameter ` FW_LND_THRTC_SC`, which reduces the TECS time constant.

### Changelog Entry
For release notes:
```
Feature: Improved altitude tracking during low height flight.
New parameter: FW_T_THR_LOW_HGT
Documentation: See `src/modules/fw_pos_control_li/FixedwingPositionControl.hpp`
```

### Testing
[VTOL test flight log](https://drive.google.com/file/d/1APXqL1cj7OxI_IxhXce8_4JQCF3uf1hY/view?usp=drive_link)

### CODE REVIEW

**Comment by RomanBapst** [comment]
@oravla5 Something that is not so nice about this is that the landing logic also sets the time contstant and it's a separate path that does not use the smoothed version. Maybe we can still look at improving this.

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp [hunk 1]
```diff
@@ -2804,6 +2816,27 @@ FixedwingPositionControl::initializeAutoLanding(const hrt_abstime &now, const po
 	}
 }
 
+bool FixedwingPositionControl::checkLowHeightConditions()
+{
+	// Are conditions for low-height
+	return _param_fw_t_thr_low_hgt.get() >= 0.f && _local_pos.dist_bottom_valid
+	       && _local_pos.dist_bottom < _param_fw_t_thr_low_hgt.get();
```
Comment (Jaeyoung-Lim) [inline src/modules/fw_pos_control/FixedwingPositionControl.cpp:2823]: I am not sure the way this is handled is desirable. This would change the TECS constant when the vehicle is flying lower than the launch altitude regardless of whether it is trying to land or navigating.

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp [hunk 2]
```diff
@@ -1236,7 +1240,9 @@ FixedwingPositionControl::control_auto_loiter(const float control_interval, cons
 		// We're in a loiter directly before a landing WP. Enable our landing configuration (flaps,
 		// landing airspeed and potentially tighter altitude control) already such that we don't
 		// have to do this switch (which can cause significant altitude errors) close to the ground.
-		_tecs.set_altitude_error_time_constant(_param_fw_thrtc_sc.get() * _param_fw_t_h_error_tc.get());
+
+		_is_low_height = true; // In low-height flight, TECS will control altitude tighter
```
Comment (Jaeyoung-Lim) [inline src/modules/fw_pos_control/FixedwingPositionControl.cpp:1244]: Please don't set this as a global variable, but try to input it as part of the method's parameters (e.g. `control_auto_loiter`) here. Same with other places that this is defined.

**Review by Jaeyoung-Lim — DISMISSED** [review]

**Comment by Jaeyoung-Lim** [comment]
@oravla5 IMHO, While TECS tuning might be part of the picture, tighter altitude tracking navigating in terrain needs to be implemented via glide slope tracking.

**Comment by RomanBapst** [comment]
@oravla5 I agree with @Jaeyoung-Lim on the point that the global state is not ideal. How about if each control_... method decides on the time constant it wants to use and then updates the slew rate itself?
Then we don't need the global state anymore and just pass down the final time consatnt to `tecs_update_pitch_throttle`.

@Jaeyoung-Lim I agree that for fixed wing landings, proper glide slope tracking is the best solution. However, here we are using a lower TECS time constant for VTOL when they are close to the ground, e..g during a VTOL landing which has nothing to do with glide slope tracking. Having a single TECS time constant makes it difficult to have tighter altitude tracking when close to terrain while also allowing a more relaxed altitude tracking at higher altitudes, which is generally desired to fly more efficiently and to calm down the throttle.

**Comment by oravla5** [comment]
@oravla5 @Jaeyoung-Lim Thanks for your feedback.

I slightly modified the implementation removing the `_is_low_height` property from the `FixedwingPositionControl` class and made it an input parameter to the `tecs_update_pitch_throttle` method. Each control mode method evaluates now the low-height requirements or enforce them (in the case of landing modes) and passes the result to the `tecs_update_ptich_throttle` method, which then updates the TECS altitude time constant accordingly.

**Comment by Jaeyoung-Lim** [comment]
> @Jaeyoung-Lim I agree that for fixed wing landings, proper glide slope tracking is the best solution. However, here we are using a lower TECS time constant for VTOL when they are close to the ground, e..g during a VTOL landing which has nothing to do with glide slope tracking. Having a single TECS time constant makes it difficult to have tighter altitude tracking when close to terrain while also allowing a more relaxed altitude tracking at higher altitudes, which is generally desired to fly more efficiently and to calm down the throttle.

@RomanBapst That makes more sense.

@oravla5 I think I am with @RomanBapst that we should have the tecs parameters as an input of `control_auto_*` methods, rather than the low_altitude boolean, since this is very specific to this use case. Rather if you specify the tecs time constant as a parameter, then you would be just configuring tecs behavior, rather than injecting mission logic.

- I still find it hard to understand why you would want the condition based on local altitude. There are a lot of use cases and deployments (especially in switerland) where the mission includes portions of the flight that flies lower than the launch position (or HOME). Given what @RomanBapst said, I would rather link the change of TECS time constant to the behavior (e.g. fw pos ctrl state) rather than relative altitude.

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp (hunk 1 above)
Comment (RomanBapst) [reply src/modules/fw_pos_control/FixedwingPositionControl.cpp:2948]: @Jaeyoung-Lim Maybe you've missed it, but this is not local altitude but distance to the ground. So this is only available when the ekf has a distance measurement to the ground. So it's completely decoupled from home altitude.

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp [hunk 3]
```diff
@@ -940,8 +944,10 @@ FixedwingPositionControl::control_auto(const float control_interval, const Vecto
 void
 FixedwingPositionControl::control_auto_fixed_bank_alt_hold(const float control_interval)
 {
-	// only control altitude and airspeed ("fixed-bank loiter")
+	// check if generic low-height flight conditions are satisfied
```
Comment (bresch) [inline src/modules/fw_pos_control/FixedwingPositionControl.cpp:947]:
```suggestion
```
I think `checkLowHeightConditions` is explicit enough, if not you should rather improve the function name than adding a comment

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp [hunk 4]
```diff
@@ -975,6 +982,9 @@ FixedwingPositionControl::control_auto_descend(const float control_interval)
 	// but not letting it drift too far away.
 	const float descend_rate = -0.5f;
 
+	// check if generic low-height flight conditions are satisfied
```
Comment (bresch) [inline src/modules/fw_pos_control/FixedwingPositionControl.cpp:985]:
```suggestion
```

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp [hunk 5]
```diff
@@ -1120,6 +1131,9 @@ FixedwingPositionControl::control_auto_position(const float control_interval, co
 
 	float yaw_body = _yaw; // yaw is not controlled, so set setpoint to current yaw
 
+	// check if generic low-height flight conditions are satisfied
```
Comment (bresch) [inline src/modules/fw_pos_control/FixedwingPositionControl.cpp:1134]:
```suggestion
```

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp [hunk 6]
```diff
@@ -1172,6 +1188,9 @@ FixedwingPositionControl::control_auto_velocity(const float control_interval, co
 
 	float yaw_body = _yaw;
 
+	// check if generic low-height flight conditions are satisfied
```
Comment (bresch) [inline src/modules/fw_pos_control/FixedwingPositionControl.cpp:1191]:
```suggestion
```

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp [hunk 7]
```diff
@@ -1242,14 +1263,20 @@ FixedwingPositionControl::control_auto_loiter(const float control_interval, cons
 	Vector2f curr_wp_local{_global_local_proj_ref.project(curr_wp(0), curr_wp(1))};
 	Vector2f vehicle_to_loiter_center{curr_wp_local - curr_pos_local};
 
+	// check if generic low-height flight conditions are satisfied
```
Comment (bresch) [inline src/modules/fw_pos_control/FixedwingPositionControl.cpp:1266]:
```suggestion
```

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp [hunk 8]
```diff
@@ -1344,6 +1372,9 @@ FixedwingPositionControl::controlAutoFigureEight(const float control_interval, c
 		tecs_fw_thr_max = _param_fw_thr_max.get();
 	}
 
+	// check if generic low-height flight conditions are satisfied
```
Comment (bresch) [inline src/modules/fw_pos_control/FixedwingPositionControl.cpp:1375]:
```suggestion
```

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp [hunk 9]
```diff
@@ -1419,6 +1450,9 @@ FixedwingPositionControl::control_auto_path(const float control_interval, const
 
 	float yaw_body = _yaw; // yaw is not controlled, so set setpoint to current yaw
 
+	// check if generic low-height flight conditions are satisfied
```
Comment (bresch) [inline src/modules/fw_pos_control/FixedwingPositionControl.cpp:1453]:
```suggestion
```

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp [hunk 10]
```diff
@@ -1465,6 +1500,9 @@ FixedwingPositionControl::control_auto_takeoff(const hrt_abstime &now, const flo
 		adjusted_min_airspeed = takeoff_airspeed;
 	}
 
+	// check if generic low-height flight conditions are satisfied
```
Comment (bresch) [inline src/modules/fw_pos_control/FixedwingPositionControl.cpp:1503]:
```suggestion
```

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp [hunk 11]
```diff
@@ -2150,6 +2197,9 @@ FixedwingPositionControl::control_manual_altitude(const float control_interval,
 		throttle_max = 0.0f;
 	}
 
+	// check if generic low-height flight conditions are satisfied
```
Comment (bresch) [inline src/modules/fw_pos_control/FixedwingPositionControl.cpp:2200]:
```suggestion
```

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp [hunk 12]
```diff
@@ -2250,6 +2301,9 @@ FixedwingPositionControl::control_manual_position(const float control_interval,
 		}
 	}
 
+	// check if generic low-height flight conditions are satisfied
```
Comment (bresch) [inline src/modules/fw_pos_control/FixedwingPositionControl.cpp:2304]:
```suggestion
```

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp [hunk 13]
```diff
@@ -2283,6 +2338,9 @@ FixedwingPositionControl::control_manual_position(const float control_interval,
 void FixedwingPositionControl::control_backtransition(const float control_interval, const Vector2f &ground_speed,
 		const position_setpoint_s &pos_sp_curr)
 {
+	// check if generic low-height flight conditions are satisfied
```
Comment (bresch) [inline src/modules/fw_pos_control/FixedwingPositionControl.cpp:2341]:
```suggestion
```

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp [hunk 14]
```diff
@@ -2756,6 +2816,9 @@ FixedwingPositionControl::tecs_update_pitch_throttle(const float control_interva
 	/* No underspeed protection in landing mode */
 	_tecs.set_detect_underspeed_enabled(!disable_underspeed_detection);
 
+	/* Update altitude time constant */
```
Comment (bresch) [inline src/modules/fw_pos_control/FixedwingPositionControl.cpp:2819]:
```suggestion
```
The function name is even more complete than the comment...

File: src/modules/fw_pos_control/FixedwingPositionControl.hpp [hunk 15]
```diff
@@ -825,6 +831,21 @@ class FixedwingPositionControl final : public ModuleBase<FixedwingPositionContro
...
+	 * @return bool True if conditions are satisfied, false otherwise
+	 */
+	bool checkLowHeightConditions();
+
+	/*
+	 * Updates TECS altitude time constant according to the is_low_height parameter.
+	 *
+	 * @param is_low_height Boolean flag defining whether we are in low-height flight
+	 * @param dt Update time step [s]
+	 */
```
Comment (bresch) [inline src/modules/fw_pos_control/FixedwingPositionControl.hpp:846]:
```suggestion
```
`void updateTECSAltitudeTimeConstant(const bool is_low_height, const float dt);` reads as "function that updates TECS altitude time constant based on is_low_height and dt" to me. The only thing relevant in this comment is maybe the unit of `dt`, but if not specified, I would anyway expect that dt is in seconds.

**Review by bresch — DISMISSED** [review]
Please, comments are not mandatory, prefer code that reads well than a comment explaining each line of code.

File: src/modules/fw_pos_control/FixedwingPositionControl.hpp (hunk 15 above)
Comment (oravla5) [reply src/modules/fw_pos_control/FixedwingPositionControl.hpp:846]: I'll keep this one to stay aligned with the rest of documented methods on this file.

File: src/modules/fw_pos_control/fw_path_navigation_params.c [hunk 16]
```diff
@@ -426,6 +426,24 @@ PARAM_DEFINE_FLOAT(FW_LND_AIRSPD, -1.f);
  */
 PARAM_DEFINE_FLOAT(FW_LND_THRTC_SC, 1.0f);
 
+/**
+ * Low-height threshold for tighter altitude tracking
+ *
+ * Defines the height (distance to bottom) threshold below which tighter altitude
+ * tracking gets enabled (see FW_LND_THRTC_SC). Below this height, TECS smoothly
+ * transitions the altitude tracking time constant from FW_T_ALT_TC to
+ * FW_LND_THRTC_SC*FW_T_ALT_TC.
```
Comment (sfuhrer) [inline src/modules/fw_pos_control/fw_path_navigation_params.c:435]: I would say here how "smooth" it is (1s per second)

File: src/modules/fw_pos_control/fw_path_navigation_params.c [hunk 17]
```diff
@@ -412,7 +412,7 @@ PARAM_DEFINE_FLOAT(FW_LND_FL_PMAX, 15.0f);
 PARAM_DEFINE_FLOAT(FW_LND_AIRSPD, -1.f);
 
 /**
- * Altitude time constant factor for landing
+ * Altitude time constant factor for landing and low-height flight
  *
  * During landing, the TECS altitude time constant (FW_T_ALT_TC)
```
Comment (sfuhrer) [inline src/modules/fw_pos_control/fw_path_navigation_params.c:417]:
```suggestion
 * The TECS altitude time constant (FW_T_ALT_TC) is multiplied by this value.
```

**Review by sfuhrer — DISMISSED** [review]
Looks okay to me, only have two minor comments.

**Review by sfuhrer — APPROVED** [review]

**Comment by sfuhrer** [comment]
We're already again out of flash on v5:` FLASH_AXIM:     2064665 B      2016 KB    100.01%`

**Review by RomanBapst — APPROVED** [review]

**Comment by RomanBapst** [comment]
@bresch @Jaeyoung-Lim Are you guys ok with the PR? If yes, please approve so that we can get it in.

**Review by Jaeyoung-Lim — COMMENTED** [review]
I agree that having different TECS time constants are useful depending on how close the terrain is.

However, I am still concerned with how this is implemented as an independent case of TECS.

I believe we have been trying hard to make the logic flow of the whole FW pos controller tractable, since nobody actually understands how many states and transitions between the states exist in the fw_pos_control module. This PR just adds yet another case that cannot be unit tested, nor is kept track in the state machine.

- Is there a reason why the condition `is_low_height` as a boolean is passed to the tecs rather than the actual time constant? Some states seem to have `is_low_height` hard coded in the states which seems unnecessary.
- Could this be implemented as a separate `FW_POSCTRL_MODE`?

**Comment by RomanBapst** [comment]
> I agree that having different TECS time constants are useful depending on how close the terrain is.
> 
> However, I am still concerned with how this is implemented as an independent case of TECS.
> 
> I believe we have been trying hard to make the logic flow of the whole FW pos controller tractable, since nobody actually understands how many states and transitions between the states exist in the fw_pos_control module. This PR just adds yet another case that cannot be unit tested, nor is kept track in the state machine.
> 
>     * Is there a reason why the condition `is_low_height` as a boolean is passed to the tecs rather than the actual time constant? Some states seem to have `is_low_height` hard coded in the states which seems unnecessary.
> 
>     * Could this be implemented as a separate `FW_POSCTRL_MODE`?

I only partially agree. I agree that it would be a bit nicer to pass the tecs time constant instead of the flag. To me it's an implementation detail that can be changed easily.
But I don't agree with the implementation being hard to trace. Initially, the implementation did actually keep a global state of the time constant and I found that hard to track. Now the implementation forces each control mode to pass the desired value for the time constant which I think makes it easy to understand. You don't need to worry about a global state that you don't know where it's updated, but you can just check in the respective control mode what the desired behavior should be. I think that is all aligned with the idea of flight tasks or flight mode manager.

**Comment by Jaeyoung-Lim** [comment]
@RomanBapst I am mainly concerned about implementation details here. I think I am convinced that the use case is valid. What would you think about defining the conditions as a `FW_STATE`, and include the condition in the state machine, such that it is clear that there are different behaviors depending on the low terrain height?

**Comment by RomanBapst** [comment]
> I am mainly concerned about implementation details here. I think I am convinced that the use case is valid. What would you think about defining the conditions as a `FW_STATE`, and include the condition in the state machine, such that it is clear that there are different behaviors depending on the low terrain height?

@Jaeyoung-Lim  I assume you mean adding a dedicated control mode for it `(FW_POSCTRL_MODE)?` 
That would mean we need to handle all the existing control modes that should have this behavior on one new control mode just because we want to pass a different time constant to TECS?
Maybe you can explain a bit more what you were thinking to make sure we are talking about the same thing.

**Comment by Jaeyoung-Lim** [comment]
@RomanBapst Yes, I am talking about extending the enums in this: https://github.com/PX4/PX4-Autopilot/blob/8f6ce4edbfd154a9f19c21d7927c865f5955c1ed/src/modules/fw_pos_control/FixedwingPositionControl.hpp#L251-L263

For me the different TECS constant close to the terrain is a different behavior, and be worth the specify it as a different state. For example, according to the implementation, the low height condition is not active for all states. Otherwise I struggle to understand when this condition is active and when it is not.

For example, I had to go through the code to understand whenthe low altitude logic is applied. The condition is NOT always based on terrain height.
- During a `FW_POSCTRL_MODE=AUTO`, the is_low_height condition is true no matter the terrain height when
  -  The current waypoint is a loiter and the next waypoint is a land waypoint.
  - If the current waypoint is a loiter but it is in a landing abort state.
- During `FW_POSCTRL_MODE=FW_POSCTRL_MODE_AUTO_LANDING_CIRCULAR` or `FW_POSCTRL_MODE=FW_POSCTRL_MODE_AUTO_LANDING_STRAIGHT`, the low height conditions seem to be always true.

I am not sure if there are more cases of this, but I had to go through the code a few times to find these conditions. Therefore I think it would be easier to understand in what cases the low altitude constraints are activated and not. 

We have a state machine in the module which is probably incomplete. If it is always based on terrain altitude, I agree we probably don't need to keep it as a state. However, there seems to be more logic to it, and I am just not comfortable on adding more complexity to the fw_pos_control logic. I doubt that anybody actually tried to draw out the flow diagram of different fw states. 

Side note: I am not sure if this behavior is valuable for manual modes, where the stick response will change based on terrain altitude, without warning the user, but maybe you have a better feeling for this.

**Comment by sfuhrer** [comment]
> For me the different TECS constant close to the terrain is a different behavior, and be worth the specify it as a different state. For example, according to the implementation, the low height condition is not active for all states. Otherwise I struggle to understand when this condition is active and when it is not.

I'm also not convinced that the TECS altitude TC variation has to be linked to a new mode. It's not black and white what a "behavioral change" is - but adjusting the weight of the controller tuning to shifting it more to tighter altitude tracking is a level below the differences the current modes have in between each other (auto vs manual, different states controlled, hard throttle and pitch limits etc).

> Side note: I am not sure if this behavior is valuable for manual modes, where the stick response will change based on terrain altitude, without warning the user, but maybe you have a better feeling for this.

I don't think the stick response changes so much, as when you deflect the stick you give a direct height rate setpoint and thus the altitude TC anyway has no effect no? And for simplicity and predictability I would make all modes the same (modes and the low altitude TC tightening feature should be considered completely orthogonal).

**Comment by RomanBapst** [comment]
@bresch Do you approve?

**Comment by bresch** [comment]
> For example, I had to go through the code to understand whenthe low altitude logic is applied. The condition is NOT always based on terrain height.
> - During a FW_POSCTRL_MODE=AUTO, the is_low_height condition is true no matter the terrain height when
>    - The current waypoint is a loiter and the next waypoint is a land waypoint.
>    - If the current waypoint is a loiter but it is in a landing abort state.
> - During FW_POSCTRL_MODE=FW_POSCTRL_MODE_AUTO_LANDING_CIRCULAR or FW_POSCTRL_MODE=FW_POSCTRL_MODE_AUTO_LANDING_STRAIGHT, the low height conditions seem to be always true.

To me that already sounds like a bug. I'm 100% in favor of what Silvan said:

> To make it cleanly decoupled form the modes: should we consider removing the hard coded is_low_height=true for the landing/early landing config phases and really only trigger it by distance to ground?

In general, if the feature is _Support tighter altitude tracking during low-height flight_ it should be kept as simple as possible and avoid the _but don't do it if the current waypoint is a loiter and the next waypoint is a land waypoint or that the current waypoint is a loiter but it is in a landing abort state_.

> there seems to be more logic to it, and I am just not comfortable on adding more complexity to the fw_pos_control logic. I doubt that anybody actually tried to draw out the flow diagram of different fw states.

I agree with that. I think that we should "modernize" the code a bit. My fear is that as it gets really hard to have all the cases in mind and that adding anything has then a high chance of breaking something else.

**Comment by RomanBapst** [comment]
@bresch 

> To me that already sounds like a bug. I'm 100% in favor of what Silvan said.

Yes, the thing is that the conditions which @Jaeyoung-Lim described are the exact same conditions that are in the code now. If we decided that we don't need them anymore because we can base everything on terrain height always, then we could remove them. However, if the distance sensor has a short range then you would get a different behavior in the sense that the TECS time constant would only be reduced at a very late stage of the landing. With the current logic, the time constant is reduced based on the waypoint type.

So in summary, the above described code is not really part of this PR and we can remove it if people are ok with it.

**Comment by RomanBapst** [comment]
@Jaeyoung-Lim As mentioned previously, the waypoint dependency in the fw land mode has been there before, and all we did was preserve that logic. If the people who care about fw landing are comfortable with removing it, then I have no problem to remove it. What would it mean in practice? If you are using a lidar that has a range of 50m, then during a land approach the TECS time constant will only be set to a tighter value once you have reached a proximity of 50m to the ground. With the current logic, the TECS time constant will be set tighter based starting the approach and not just on the proximity to terrain. So even if your land approach started at 300m above ground, you would have tighter altitude tracking from the beginning, because they waypoint type tells you that you are doing an approach.
